src/Microsoft.Extensions.Http.Polly/PolicyHttpMessageHandler.cs

```diff
@@ -77,7 +82,38 @@ protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage reques
                 throw new ArgumentNullException(nameof(request));
             }
 
-            return _policy.ExecuteAsync((ct) => base.SendAsync(request, ct), cancellationToken);
+            // Guarantee the existance of a context for every policy execution, but only create a new one if needed. This
+            // allows later handlers to flow state if desired.
+            var context = request.GetPollyContext();
+            if (context == null)
+            {
+                context = new Context(Guid.NewGuid().ToString());
+                request.SetPollyContext(context);
+            }
+
+            return _policy.ExecuteAsync((c, ct) => SendCoreAsync(request, c, ct), context, cancellationToken);
```

Comment:
Note that I'm explicitly **not** cleaning up the context here. It's not clear that there's any value to that, but I'm interested in thoughts.

I have a suspicion that we might want to... but it would only affect some truly weird scenarios. Like if someone has implemented their own retry support, but they aren't setting the context themself.

Reply:
>explicitly not cleaning up the context here

It seems correct (in fact important) that we don't clean up the context.  This line in execution could be hit within the looping flowing from an enclosing retry policy.  We'd want the same context to hold through all retries, so we shouldn't (for example) set it to null here.  (That iiuc your thought re 'clean up context' - come back on this if I'm missing something, or if you see scenarios that might contradict that!)

Reply:
Ok good, sounds like this is right then.